src/installer.ts

```diff
@@ -185,15 +201,44 @@ async function installGoVersion(
     extPath = path.join(extPath, 'go');
   }
 
-  core.info('Adding to the cache ...');
-  const cachedDir = await tc.cacheDir(
-    extPath,
-    'go',
-    makeSemver(info.resolvedVersion),
-    arch
+  // for github hosted windows runner handle latency of OS drive
+  // by avoiding write operations to C:
+
+  if (!isWindows) return addExecutablesToCache(extPath, info, arch);
+
+  const isHosted =
+    process.env['RUNNER_ENVIRONMENT'] === 'github-hosted' ||
+    process.env['AGENT_ISSELFHOSTED'] === '0';
+  if (!isHosted) return addExecutablesToCache(extPath, info, arch);
+
+  const defaultToolCacheRoot = process.env['RUNNER_TOOL_CACHE'];
+  if (!defaultToolCacheRoot) return addExecutablesToCache(extPath, info, arch);
+
+  if (!fs.existsSync('d:\\') || !fs.existsSync('c:\\'))
+    return addExecutablesToCache(extPath, info, arch);
+
+  const substitutedToolCacheRoot = defaultToolCacheRoot
+    .replace('C:', 'D:')
+    .replace('c:', 'd:');
+  // make toolcache root to be on drive d:
+  process.env['RUNNER_TOOL_CACHE'] = substitutedToolCacheRoot;
+
+  const actualToolCacheDir = await addExecutablesToCache(extPath, info, arch);
+
+  // create a link from c: to d:
+  const defaultToolCacheDir = actualToolCacheDir.replace(
+    substitutedToolCacheRoot,
+    defaultToolCacheRoot
   );
-  core.info(`Successfully cached go to ${cachedDir}`);
-  return cachedDir;
+  fs.mkdirSync(path.dirname(defaultToolCacheDir), {recursive: true});
+  fs.symlinkSync(actualToolCacheDir, defaultToolCacheDir, 'junction');
+  core.info(`Created link ${defaultToolCacheDir} => ${actualToolCacheDir}`);
+
+  // restore toolcache root to default drive c:
+  process.env['RUNNER_TOOL_CACHE'] = defaultToolCacheRoot;
+
+  // make outer code to continue using toolcache as if it were installed on c:
+  return defaultToolCacheDir;
```

Comment:
I'd suggest to incapsulate this code into the wrapper function, which inside of it will decide how to save go into the toolcache.